pkg/controller/statefulset/stateful_set_utils.go

```diff
@@ -253,6 +253,30 @@ func newVersionedStatefulSetPod(currentSet, updateSet *appsv1alpha1.StatefulSet,
 	return pod
 }
 
+// isCurrentRevisionNeeded calculate if the 'ordinal' Pod should be current revision.
+func isCurrentRevisionNeeded(set *appsv1alpha1.StatefulSet, updateRevision string, ordinal int, replicas []*v1.Pod) bool {
+	if set.Spec.UpdateStrategy.Type != apps.RollingUpdateStatefulSetStrategyType {
+		return false
+	}
+	if set.Spec.UpdateStrategy.RollingUpdate == nil {
+		return ordinal < int(set.Status.CurrentReplicas)
+	}
+	if set.Spec.UpdateStrategy.RollingUpdate.UnorderedUpdate == nil {
+		return ordinal < int(*set.Spec.UpdateStrategy.RollingUpdate.Partition)
+	}
+
+	var noUpdatedReplicas int
+	for i, pod := range replicas {
```

Comment:
Does it make sense to keep a counter array in status which counts the number of pods in each version? This may help avoid looping all Pods over and over.

Reply:
Status is an unreliable data, it is calculated by controller asynchronously. Besides, a simple loop is not a big deal.